add sessionDidLoadWebView to TurboNavigationDelegate

yanshiyason opened the pull request:
Hi @joemasilotti, 

I am implementing some functionality which requires me to set the navigation delegate of the web view.
According to the Turbo docs, the way to do this is to implement the `sessionDidLoadWebView` method.

https://github.com/hotwired/turbo-ios/blob/main/Docs/Overview.md#becoming-the-web-views-navigation-delegate

> When doing the cold boot visit, Turbo will need to take over as the WKWebView’s navigationDelegate to know when the page loads. After that, it's likely your application will want to become the navigationDelegate so you can have full control over the various methods.

I added `sessionDidLoadWebView` to the `TurboNavigationDelegate` in my fork, and I would like to know if this is something you would merge into `TurboNavigator`.

If these changes are not suitable, please let me know if there is an alternative way of becoming the web view's navigation delegate.

Thank you!

yanshiyason committed:
add sessionDidLoadWebView to TurboNavigationDelegate

joemasilotti commented:
Thanks for this, @yanshiyason. Out of curiosity, why do you need to be the delegate of the web view?

Regardless, I think something like this definitely needs to be exposed to ensure backwards compatibility with turbo-ios. One thing I'm not 100% sure on is if I want to expose the concept of the `Session` at all.

Your PR is in line with the other methods above and below it. I'm thinking more long term about the project's surface area.

yanshiyason commented:
@joemasilotti Thank you for your quick response.

> Out of curiosity, why do you need to be the delegate of the web view?

I am adding a native file download functionality. When clicking on a link with a "download" attribute (`<a href=/my.pfd download>`) we want to show a `UIDocumentPickerViewController` and allow the user to save the file.

If the pdf link doesn't have the "download" attribute, then it will open in a `SFSafariViewController` (based on the file extension), the user can then view the PDF as usual, and download it from the page options.

If the pdf link has the "download" attribute, and you click it, nothing seems to happen at first, and that is because you must implement the `WKDownloadDelegate` interface. (And to become the download delegate, you must first become the navigation delegate)

I thought it would be possible to detect the "disposition=attachment" parameter in the URL, inside the `openExternalURL` method, and show the document picker there, but the problem is that I don't seem to have access to the filename, and I am not sure the download request will be authenticated since it's performed outside the web view 🤔 (only problematic if the download endpoint is authenticated)

Please let me know if you know alternative ways of implementing this kind of functionality.

Thank you!

joemasilotti committed:
Clean up documentation

joemasilotti commented:
Sorry for the delay on this! It took me a while to wrap my head around the situation.

joemasilotti committed:
Ensure the web view's navigation delegate is set

Otherwise, external URLs aren't routed correctly. The developer can
override this by setting a new delegate in their implementation.
